File: Hackerrank/One Month Prep/Python/HR3-Sherlock and the Valid String.py
#!/bin/python3

import math
import os
import random
import re
import sys

#
# Complete the 'isValid' function below.
#
# The function is expected to return a STRING.
# The function accepts STRING s as parameter.
#

def isValid(s):
    if len(s)<2:
        return "YES";
    
    d={}
    for i in s:
        if i in d.keys():
            d[i]+=1
        else:
            d[i]=1
    print(d)
    
    v={}
    for i in d.values():
        if i in v.keys():
            v[i]+=1
        else:
            v[i]=1
    v=dict(sorted(v.items()))
    print(v)
    if len(v)<3:
        if len(v)==1:
            return "YES"
        
        if 1 in v.keys():
            if v[1]==1:
                return "YES"
        
        k=list(v.keys())
        print(k,v[k[1]],v[k[0]])
        
        if k[1]-k[0]==1 and v[k[1]]==1:
            return "YES"
            
    
    return "NO"

if __name__ == '__main__':
    fptr = open(os.environ['OUTPUT_PATH'], 'w')

    s = input()

    result = isValid(s)

    fptr.write(result + '\n')

    fptr.close()

File: Hackerrank/One Month Prep/CPP/HR3-Sherlock and the Valid String.cpp
#include <bits/stdc++.h>

using namespace std;

/*
 * Complete the 'isValid' function below.
 *
 * The function is expected to return a STRING.
 * The function accepts STRING s as parameter.
 */

string isValid(string s) 
{
    if (s.size()==1 || s.size()==2)
        return "YES";
    
    vector<int> q('z'-'a'+1,0);
    
    for(auto i: s)
        q[i-'a']++;
    
    sort(q.begin(),q.end());
    
    vector<vector<int>>m;
    int c=0;
    int y=0;
    for( auto i: q)
    {
        if(i!=0)
        {
            y=0;c=0;
            while(y<m.size())
            {
                if(i==m[y][0])
                {
                    c=1;
                    m[y][1]++;
                    break;
                }
                y++;
            }
            if(c==0)
                m.push_back({i,1});
        }
    }
    
    for( auto i: m)
    {
        cout<<i[0]<<" "<<i[1]<<endl;
    }
    
    if(m.size()==1)
        return "YES";
    
    if(m.size()<3)
    {
        
        if(m[0][0]==1 && m[0][1]==1)
            return "YES";
        
        if(m[1][1]==1 && m[1][0]-m[0][0]==1)
            return "YES";
    }
        
    return "NO";
    
    
}

int main()
{
    ofstream fout(getenv("OUTPUT_PATH"));

    string s;
    getline(cin, s);

    string result = isValid(s);

    fout << result << "\n";

    fout.close();

    return 0;
}

